src/Data/Variant.purs

```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
+  where
+    variantBoundedEnums _ = L.Cons dict (variantBoundedEnums (RLProxy ∷ RLProxy rs))
+      where
+      dict ∷ BoundedEnumDict VariantCase
+      dict =
+        { pred: coerceAToMbA pred
+        , succ: coerceAToMbA succ
+        , fromEnum: coerceFromEnum fromEnum
+        , toEnum: coerceToEnum toEnum
+        , cardinality: coerceCardinality cardinality
+        }
+
+      coerceA ∷ a → VariantCase
+      coerceA = unsafeCoerce
+
+      coerceAToMbA ∷ (a → M.Maybe a) → VariantCase → M.Maybe VariantCase
+      coerceAToMbA = unsafeCoerce
+
+      coerceFromEnum ∷ (a → Int) → VariantCase → Int
+      coerceFromEnum = unsafeCoerce
+
+      coerceToEnum ∷ (Int → M.Maybe a) → Int → M.Maybe VariantCase
+      coerceToEnum = unsafeCoerce
+
+      coerceCardinality ∷ Cardinality a → Int
+      coerceCardinality = unsafeCoerce
+
+instance enumVariant
+  ∷ ( R.RowToList r rl
+    , VariantTags rl
+    , VariantEqs rl
+    , VariantOrds rl
+    , VariantBoundedEnums rl
+    )
+  ⇒ Enum (Variant r) where
+    pred a = coerceOut case dict.pred (unsafeCoerce arep.value) of
+      M.Just b → M.Just $ VariantRep arep{ value = b }
+      M.Nothing
+        | tagIx == 0 → M.Nothing
+        | otherwise →
+          map (\p → VariantRep { type: T.fst p, value: (T.fst $ T.snd p).top })
+          $ L.index zipped $ tagIx - 1
+        where
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBoundedEnums prx
+        bounds = variantBounded prx
+        zipped = L.zip tags $ L.zip bounds dicts
+
+        coerceOut ∷ M.Maybe (VariantRep VariantCase) → M.Maybe (Variant r)
+        coerceOut = unsafeCoerce
+
+        VariantRep arep = unsafeCoerce a
+
+        tagIx ∷ Int
+        tagIx = fromJust "enum" $ L.findIndex (\(T.Tuple s _) → arep.type == s) zipped
+
+        pack = fromJust "enum" $ L.index zipped tagIx
+
+        dict ∷ BoundedEnumDict VariantCase
+        dict = T.snd $ T.snd pack
+
+    succ a = coerceOut case dict.succ (unsafeCoerce arep.value) of
+      M.Just b → M.Just $ VariantRep arep{ value = b }
+      M.Nothing
+        | tagIx == totalTags - 1 → M.Nothing
+        | otherwise →
+          map (\p → VariantRep { type: T.fst p, value: (T.fst $ T.snd p).bottom })
+          $ L.index zipped $ tagIx + 1
+        where
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBoundedEnums prx
+        bounds = variantBounded prx
+        zipped = L.zip tags $ L.zip bounds dicts
+
+        coerceOut ∷ M.Maybe (VariantRep VariantCase) → M.Maybe (Variant r)
+        coerceOut = unsafeCoerce
+
+        VariantRep arep = unsafeCoerce a
+
+        tagIx ∷ Int
+        tagIx = fromJust "enum" $ L.findIndex (\(T.Tuple s _) → arep.type == s) zipped
+
+        pack = fromJust "enum" $ L.index zipped tagIx
+
+        dict ∷ BoundedEnumDict VariantCase
+        dict = T.snd $ T.snd pack
+
+        totalTags = L.length tags
```

Comment:
Same as `pred`, except I think this performs an additional traversal for length.